sdk/search/azure-search/src/main/java/com/azure/search/SearchIndexAsyncClient.java

```diff
@@ -0,0 +1,736 @@
+import com.azure.search.models.AutocompleteItem;
+import com.azure.search.models.AutocompleteOptions;
+import com.azure.search.models.AutocompleteRequest;
+import com.azure.search.models.IndexAction;
+import com.azure.search.implementation.models.IndexActionType;
+import com.azure.search.models.IndexBatch;
+import com.azure.search.models.IndexDocumentsResult;
+import com.azure.search.models.RequestOptions;
+import com.azure.search.models.SearchOptions;
+import com.azure.search.models.SearchRequest;
+import com.azure.search.models.SearchResult;
+import com.azure.search.models.SuggestOptions;
+import com.azure.search.models.SuggestRequest;
+import com.azure.search.models.SuggestResult;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.commons.lang3.StringUtils;
+import reactor.core.publisher.Mono;
+
+import java.util.List;
+import java.util.function.Function;
+
+import static com.azure.core.util.FluxUtil.withContext;
+
+/**
+ * Cognitive Search Asynchronous Client to query an index and upload, merge, or delete documents
+ */
+@ServiceClient(builder = SearchIndexClientBuilder.class, isAsync = true)
+public class SearchIndexAsyncClient {
+
+    /**
+     * The lazily-created serializer for search index client.
+     */
+    private static final SerializerAdapter SERIALIZER = initializeSerializerAdapter();
+
+    /**
+     * Search REST API Version
+     */
+    private final SearchServiceVersion apiVersion;
+
+    /**
+     * The endpoint for the Azure Cognitive Search service.
+     */
+    private final String endpoint;
+
+    /**
+     * The name of the Azure Cognitive Search index.
+     */
+    private final String indexName;
+
+    /**
+     * The logger to be used
+     */
+    private final ClientLogger logger = new ClientLogger(SearchIndexAsyncClient.class);
+
+    /**
+     * The underlying AutoRest client used to interact with the Azure Cognitive Search service
+     */
+    private final SearchIndexRestClientImpl restClient;
+
+    /**
+     * The pipeline that powers this client.
+     */
+    private final HttpPipeline httpPipeline;
+
+    /**
+     * Package private constructor to be used by {@link SearchIndexClientBuilder}
+     */
+    SearchIndexAsyncClient(String endpoint, String indexName, SearchServiceVersion apiVersion,
+        HttpPipeline httpPipeline) {
+
+        SearchServiceUrlParts parts = SearchServiceUrlParser.parseServiceUrlParts(endpoint);
+
+        if (StringUtils.isBlank(indexName)) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("Invalid indexName"));
+        }
+        if (apiVersion == null) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("Invalid apiVersion"));
+        }
+        if (httpPipeline == null) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("Invalid httpPipeline"));
+        }
```

Comment:
Should these checks be made in the client builder instead of the constructor?

Reply:
added validation to builder as well